### PR TITLE
Issue 24: Build with ASAN and check leaks

### LINKED ISSUE #9: Proper reference counting
The Z3 C API uses manual reference counting. We currently do not care about reference counting, which likely results in memory leaks. To provide a usable API (without manual reference counting in Ada) we might have to use controlled objects for most of the types.

### LINKED ISSUE #24: Build with ASAN and check leaks
Build z3 and our test suite with ASAN and enable resource leak checks:

```
   package Builder is                                                                                                   
      for Global_Compilation_Switches ("Ada") use ("-fsanitize=address", "-fno-omit-frame-pointer");      
   end Builder; 

   package Linker is                                                                                                    
      for Default_Switches ("Ada") use ("-lasan");                                                            
   end Linker; 
```

### PR DESCRIPTION
**Pull request by jklmnn**
Closes #9
Closes #24

### CODE REVIEW

File: tests/z3-tests.adb [hunk 1]
```diff
@@ -813,6 +813,46 @@ package body Z3.Tests is
 
    ---------------------------------------------------------------------------
 
+   procedure Test_Reference_Counting (T : in out Test_Cases.Test_Case'Class)
+   is
+      pragma Unreferenced (T);
```
Comment (senier) [inline tests/z3-tests.adb:818]: Nit: In other places I started using `Unused_T` as a parameter name which avoids the additional pragma.

File: tests/z3-tests.adb [hunk 2]
```diff
@@ -813,6 +813,46 @@ package body Z3.Tests is
...
+      is
+         Optimize2 : Z3.Optimize := Create (Context);
+      begin
+         Optimize2 := Local_Optimize;
+         pragma Inspection_Point (Optimize2);
+      end Copy_Optimize;
+   begin
+      Copy_Ctx (Context);
+      Copy_Solver (Solver);
+      Copy_Optimize (Optimize);
```
Comment (senier) [inline tests/z3-tests.adb:851]: What is being tested here (I don't see any assertion)?

File: z3.ads [hunk 3]
```diff
@@ -434,15 +434,22 @@ package Z3 is  --  GCOV_EXCL_LINE
...
+   type Reference_Counter is access Natural;
 
-   type Context is
-      record
-         Data : z3_api_h.Z3_context;
-      end record;
+   type Context is new Ada.Finalization.Controlled with record
+      Data      : z3_api_h.Z3_context;
+      Config    : z3_api_h.Z3_config;
+      Ref_Count : Reference_Counter;
```
Comment (senier) [inline z3.ads:442]: What is the reason for `Ref_Count` being an access type?

**Review by senier — CHANGES_REQUESTED** [review]

File: tests/z3-tests.adb (hunk 2 above)
Comment (jklmnn) [reply tests/z3-tests.adb:851]: This tests the reference counting by creating copies in procedures and letting them go out of scope. The test is checked by the address sanitizer. We don't need AUnit for this test but it doesn't make sense to create a separate test binary just for that.

File: z3.ads (hunk 3 above)
Comment (jklmnn) [reply z3.ads:442]: When `Context` is copied, `Adjust` will be called on the copy. Due to the access type all copies hold a reference on the same reference counter. If it was just an integer all copies would contain their own reference counter and there would be no way to keep them synchronized.

Comment (senier) [reply z3.ads:442]: I see.

File: tests/z3-tests.adb (hunk 2 above)
Comment (senier) [reply tests/z3-tests.adb:851]: I don't see a problem with having it here, but we should add a comment to remind our future selves of that.

File: Makefile [hunk 4]
```diff
@@ -9,21 +10,22 @@ GCOVR_OPTS = \
...
 z3/z3/build/libz3.so: z3/z3/build/Makefile
-	$(MAKE) -C z3/z3/build
+	make -C z3/z3/build -j$(shell nproc)
 
 build: z3/z3/build/libz3.so
 	@$(GPRBUILD) $(GPRBUILD_OPTS) -P tests/tests
 
 clean:
 	gnatclean -P tests/tests.gpr
+	gprclean -P tests/aunit/lib/gnat/aunit.gpr
```
[the remark this answers is not in the document]
Comment (jklmnn) [reply Makefile:27]: I'm asking myself why we would use `gprbuild` through a variable? Do we have any other program that would replace `gprbuild`?

[the remark this answers is not in the document]
Comment (senier) [reply Makefile:27]: I guess I had some reason to do that, but it doesn't make a lot of sense (when using a different toolchain you'll normally adjust PATH which gives you the correct gprbuild). I suggest you remove it.

File: tests/z3-tests.adb (hunk 1 above)
Comment (jklmnn) [reply tests/z3-tests.adb]: Fixed in https://github.com/Componolit/AZ3/pull/35/commits/2aac0b6cc800384affd2daebbe6e61db6008163c.

File: tests/z3-tests.adb (hunk 2 above)
Comment (jklmnn) [reply tests/z3-tests.adb:833]: Fixed by https://github.com/Componolit/AZ3/pull/35/commits/476de4c4fd0e949efae48fc1b63a428fbef58423.

File: Makefile (hunk 4 above)
[the remark this answers is not in the document]
Comment (jklmnn) [reply Makefile]: Fixed by https://github.com/Componolit/AZ3/pull/35/commits/3f4fe80be9e35fb2b6768d2283e1f824b0450fa5.

**Review by senier — APPROVED** [review]
Please don't forget to associate the PRs with their respective ticket(s).

**Pull request merged**